### ISSUE TITLE
LoadError: cannot load such file -- tiny_tds/tiny_tds

### ISSUE THREAD

**Issue by aharpervc** [rails-sqlserver/tiny_tds#391]
## Before submitting an issue please check these first!

* On Windows? If so, do you need devkit for your ruby install?
  - Yes
  - Yes? I don't exactly understand the question, but I do have devkit installed
* ~Using Ubuntu? If so, you may have forgotten to install FreeTDS first.~
  - N/A
* Are you using FreeTDS 0.95.80 or later? Check `$ tsql -C` to find out.
  - Unclear, `tsql -C` does not produce any output, but the folder path that `tsql` in implies freetds 1.00.27
* ~If not, please update then uninstall the TinyTDS gem and re-install it.~
  - N/A
* ~Have you made sure to [enable SQL Server authentication](http://bit.ly/1Kw3set)?~
  - N/A
* Doing work with threads and the raw client? Use the ConnectionPool gem?
  - No

If none of these help. Please fill out the following:

## Environment

**Operating System**

Please describe your operating system and version here.

  - Windows 10 x64 v1709 (16299.19)

**TinyTDS Version and Information**

```
Please paste the full output of `ttds-tsql -C` (or `tsql -C` for older versions
of TinyTDS) here. If TinyTDS does not install, please provide the gem version.
```
```
$ ttds-tsql -C
ttds-tsql : The term 'ttds-tsql' is not recognized as the name of a cmdlet, function, script file, or operable program
```
```
$ tsql -C
[TinyTds][v1.0.5][tsql]: c:/Program Files (x86)/ruby/233/lib/ruby/gems/2.3.0/gems/tiny_tds-1.0.5-x86-mingw32/exe/tsql.EXE
```

**FreeTDS Version**

Please provide your system's FreeTDS version. If you are using the pre-compiled
windows gem you may omit this section.

- I am using the Windows gem

## Description

Please describe the bug or feature request here.

I am not able to use TinyTds 2.1.0 on Ruby 2.3.3. The gem appears to install fine, but when I try to require it, I get the error `LoadError: cannot load such file -- tiny_tds/tiny_tds`.

It looks like this, followed by more stack trace:

```
$ ruby -v
ruby 2.3.3p222 (2016-11-21 revision 56859) [i386-mingw32]

$ gem install tiny_tds
Fetching: tiny_tds-2.1.0-x86-mingw32.gem (100%)
Successfully installed tiny_tds-2.1.0-x86-mingw32
1 gem installed

$ irb
irb(main):001:0> require 'tiny_tds'
LoadError: cannot load such file -- tiny_tds/tiny_tds
        from C:/bin/ruby/233/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
```

Since we have apparently not dropped support for ruby 2.3.3, I would like to figure out what is going on.

**Comment by coderjoe** [comment]
@metaskills I'm at the point where I feel like we do need to automatically enable devkit in the gem. If @aharpervc (who knows the project very well) can't make it work without guidance, I don't think we can expect the average random windows ruby developer to figure it out either.

I'm very welling to review PRs to fix this as well as give the PR a try if it sticks around until next weekend.

**Comment by aharpervc** [comment]
I can hack it for myself with `require devkit`, as mentioned in the last thread, but since that isn't good/safe, I'm out of ideas.

**Comment by coderjoe** [comment]
Until a solution is created, you need to `require 'tiny_tds'` in your rails app before the rails app is bootstrapped and therefore before tiny_tds is required.

**Comment by aharpervc** [comment]
I can deal with that. However, it's still unclear to me what devkit is changing to make library loading work. It's changing the ENV `PATH` right? So some feature of devkit resides in it's own folder, added to the PATH, and now library loading is working. Is there a way to discover whatever that is, and load it selectively, without changing the PATH?

**Comment by coderjoe** [comment]
@aharpervc unfortunately no. Because of how devkit works its dependent on the user's choice of installation directory. That's why the devkit shim exists, to modify a ton of environment variables all at once if your project needs to use devkit.

If you'd like to learn more I'm happy to discuss it, but I'd rather do that via gitter than in this issue.

**Comment by mbratch** [comment]
@aharpervc  where did you put `require 'devkit'` to get it to work? I tried it as the first line in `config/application.rb` and `config/boot.rb` and I still see the TDS loading error.

**Comment by mbratch** [comment]
@aharpervc and @coderjoe thanks for the tip on using `require 'devkit'` in the application. I put it just before `Bundler.require(*Rails.groups)` in my `config/application.rb` and it seems to work around this issue.

**Comment by aharpervc** [comment]
I had been hacking tiny_tds.rb, before it tries to load the native library, but hypothetically any time before `require 'tiny_tds'` should work.

**Comment by wakeless** [comment]
Is there a fix for this, we are using OCRA to create a binary and this doesn't seem to work with the `require 'devkit'` fix.

**Comment by JohnAtFenestra** [comment]
Adding to this issue some additional information.

I am encountering this issue on Windows 7 using straight Ruby. 

```
>ruby --version
ruby 2.4.4p296 (2018-03-28 revision 63013) [i386-mingw32]
>gem list tiny_tds

*** LOCAL GEMS ***

tiny_tds (2.1.2 x86-mingw32)
```
I added a tiny piece of logging to tiny_tds.rb to report the attempted load version and to report the exception message in the rescue. 

```
# Temporary add bin directories for DLL search, so that freetds DLLs can be found.
  add_dll_paths.call( TinyTds::Gem.ports_bin_paths ) do
    begin
	  puts "tiny_tds/#{ver}/tiny_tds"
      require "tiny_tds/#{ver}/tiny_tds"
    rescue LoadError => boom
	  puts boom.message
      require 'tiny_tds/tiny_tds'
    end
  end
```

Here are the results of IRB. Note that I explicitly require `devkit` before requiring `tiny_tds`:

```
>irb                                                                                                                      
irb(main):001:0> require 'devkit'                                                                                                              
=> true                                                                                                                                        
irb(main):002:0> require 'tiny_tds'                                                                                                            
tiny_tds/2.4/tiny_tds                                                                                                                          
126: The specified module could not be found.   - C:/Ruby24/lib/ruby/gems/2.4.0/gems/tiny_tds-2.1.2-x86-mingw32/lib/tiny_tds/2.4/tiny_tds.so   
tiny_tds/2.4/tiny_tds                                                                                                                          
126: The specified module could not be found.   - C:/Ruby24/lib/ruby/gems/2.4.0/gems/tiny_tds-2.1.2-x86-mingw32/lib/tiny_tds/2.4/tiny_tds.so   
tiny_tds/2.4/tiny_tds                                                                                                                          
126: The specified module could not be found.   - C:/Ruby24/lib/ruby/gems/2.4.0/gems/tiny_tds-2.1.2-x86-mingw32/lib/tiny_tds/2.4/tiny_tds.so   
tiny_tds/2.4/tiny_tds                                                                                                                          
126: The specified module could not be found.   - C:/Ruby24/lib/ruby/gems/2.4.0/gems/tiny_tds-2.1.2-x86-mingw32/lib/tiny_tds/2.4/tiny_tds.so   
LoadError: cannot load such file -- tiny_tds/tiny_tds                                                                                          
        from C:/Ruby24/lib/ruby/2.4.0/rubygems/core_ext/kernel_require.rb:55:in `require'                                                      
        from C:/Ruby24/lib/ruby/2.4.0/rubygems/core_ext/kernel_require.rb:55:in `require'                                                      
        from C:/Ruby24/lib/ruby/gems/2.4.0/gems/tiny_tds-2.1.2-x86-mingw32/lib/tiny_tds.rb:48:in `rescue in block in <top (required)>'         
        from C:/Ruby24/lib/ruby/gems/2.4.0/gems/tiny_tds-2.1.2-x86-mingw32/lib/tiny_tds.rb:43:in `block in <top (required)>'                   
        from C:/Ruby24/lib/ruby/gems/2.4.0/gems/tiny_tds-2.1.2-x86-mingw32/lib/tiny_tds.rb:37:in `block in <top (required)>'                   
        from C:/Ruby24/lib/ruby/gems/2.4.0/gems/tiny_tds-2.1.2-x86-mingw32/lib/tiny_tds.rb:34:in `block (2 levels) in <top (required)>'        
        from C:/Ruby24/lib/ruby/gems/2.4.0/gems/tiny_tds-2.1.2-x86-mingw32/lib/tiny_tds.rb:24:in `rescue in block in <top (required)>'         
        from C:/Ruby24/lib/ruby/gems/2.4.0/gems/tiny_tds-2.1.2-x86-mingw32/lib/tiny_tds.rb:17:in `block in <top (required)>'                   
        from C:/Ruby24/lib/ruby/gems/2.4.0/gems/tiny_tds-2.1.2-x86-mingw32/lib/tiny_tds.rb:33:in `block in <top (required)>'                   
        from C:/Ruby24/lib/ruby/gems/2.4.0/gems/tiny_tds-2.1.2-x86-mingw32/lib/tiny_tds.rb:42:in `<top (required)>'                            
        from C:/Ruby24/lib/ruby/2.4.0/rubygems/core_ext/kernel_require.rb:133:in `require'                                                     
        from C:/Ruby24/lib/ruby/2.4.0/rubygems/core_ext/kernel_require.rb:133:in `rescue in require'                                           
        from C:/Ruby24/lib/ruby/2.4.0/rubygems/core_ext/kernel_require.rb:39:in `require'                                                      
        from (irb):2                                                                                                                           
        from C:/Ruby24/bin/irb.cmd:19:in `<main>'                                                                                              
irb(main):003:0>
```


I have confirmed that `tiny_tds.so` exists at the specified location:

```
>dir C:\Ruby24\lib\ruby\gems\2.4.0\gems\tiny_tds-2.1.2-x86-mingw32\lib\tiny_tds\2.4
 Volume in drive C has no label.
 Volume Serial Number is 5C90-F2CA

 Directory of C:\Ruby24\lib\ruby\gems\2.4.0\gems\tiny_tds-2.1.2-x86-mingw32\lib\tiny_tds\2.4

09/24/2018  10:11 AM    <DIR>          .
09/24/2018  10:11 AM    <DIR>          ..
09/24/2018  10:11 AM           320,622 tiny_tds.so
               1 File(s)        320,622 bytes
```

Here is an extraction of my path. To eliminate noise, I reduced the included list to only those items containing `/ruby/i` in the path:

```
[                                                                                                                                
    "C:\\Ruby24\\bin",                                                                                                           
    "C:\\Ruby24\\lib\\ruby\\gems\\2.4.0\\gems\\tiny_tds-2.1.2-x86-mingw32\\ports\\i686-w64-mingw32\\freetds\\1.00.27\\bin",      
    "C:\\Ruby24\\lib\\ruby\\gems\\2.4.0\\gems\\tiny_tds-2.1.2-x86-mingw32\\ports\\i686-w64-mingw32\\openssl\\1.1.0e\\bin",       
    "c:\\Ruby24\\msys32\\mingw32\\bin"                                                                                           
]
```

I'm out of ideas of what to look at next.

**Comment by JohnAtFenestra** [comment]
Adding a comment. You may want to look at my findings for Ruby 2.4 + on https://github.com/rails-sqlserver/tiny_tds/issues/416. Try reinstalling the gem with:

```
gem install --platform ruby tiny_tds
```

**Comment by meric426** [comment]
> Is there a fix for this, we are using OCRA to create a binary and this doesn't seem to work with the `require 'devkit'` fix.

Did you ever solve this?

**Comment by JohnAtFenestra** [comment]
@meric426 

See my comment directly above yours. #416 documents how using `--platform ruby` corrects the issue.

**Comment by meric426** [comment]
> @meric426
> 
> See my comment directly above yours. #416 documents how using `--platform ruby` corrects the issue.

@JohnAtFenestra I can get the script running. But trying to package it into a `.exe` with OCRA doesn't work

**Comment by JohnAtFenestra** [comment]
If tiny_tds runs from a script but not after packaging, then I'd say you have an issue with Ocra packaging rules. Without seeing your error, my guess is that Ocra isn't dealing with the issues that `--platform ruby` resolves. If you look at my notes on #416, you'll see that the .so files existed in the correct locations, however, for whatever reason, tiny_tds still didn't work. Building the extensions with `--platform ruby` resolved them, but I didn't dig deeper into why.

Good luck!

**Comment by vidhijain24** [comment]
Hi Team,

I am facing same error message when I require 'tiny_tds' in my code.

My ruby version is Ruby30-x64.  Also I have installed tiny_tds using below command:
gem install tiny_tds -v '2.0.0'

Please help me to resolve my issue.Thanks

**Comment by swistak35** [comment]
@meric426 Have you managed to solve your issue with tiny_tds and ocra?

**Comment by meric426** [comment]
> @meric426 Have you managed to solve your issue with tiny_tds and ocra?

Nah, abandoned that project a long time ago

**Comment by swistak35** [comment]
Thank you for your response.

Would one of the maintainers be willing to provide **paid** support for that issue? (I'd have to ask about the details, but it would be possible) If the setup part sound too intimidating (Windows, I know...), we could also pair-program the issue on my machine. Feel free to contact me through the e-mail on my github profile.

**Comment by swistak35** [comment]
I've figured this out. The problem was missing `dlls`, in the previous tiny_tds version I've used (1.3.0) they were included, in 2.1.5 they were not and ocra didn't include them. That should be I guess expected (not ocra's fault), it's just that the error message is cryptic and doesn't provide good entrypoint for debugging.

I've figured out which libraries were missing by using procmon tool as described here: https://github.com/larsch/ocra/issues/174#issuecomment-1112104881

It turned out I was missing following dlls:
- libiconv-2.dll
- libsybdb-5.dll
- libcrypto32-1.1.0e-i686-w64-mingw32.dll
- libssl32-1.1.0e-i686-w64-mingw32.dll

The first one (libiconv) is included in ruby 2.6 installation. The rest of them I took from `tiny_tds` 1.3.0 gem (`ports` directory). The ones from tiny_tds 1.3.0 I've put in my ruby 2.6 bindir (`C:/Ruby26/bin`), and I've added following options to ocra command:

```
--dll ruby_builtin_dlls/libiconv-2.dll --dll libsybdb-5.dll --dll libcrypto32-1.1.0e-i686-w64-mingw32.dll --dll libssl32-1.1.0e-i686-w64-mingw32.dll
```

**Comment by andyundso** [comment]
I'll close this issue.

`require devkit` should no longer be necessary on Windows since `RubyInstaller2` works differently internally.

Second, we merged a PR (#544) which was released in v2.1.6 that fixes installation on Windows without using the precompiled gem, including testing on CI to make sure it does not break.

If there are still issues on Windows or specifically with `ocra` / `ocran`, feel free to open a new issue.
